### PR TITLE
TM-87: Update vocabs

### PR DESCRIPTION
**Pull request by ManuelTS**
Closes TM-87.

### CODE REVIEW

File: be/src/controllers/rest.router.ts
Comment (ja-fra) [inline be/src/controllers/rest.router.ts]: Since we will need this pattern at several places, please move it into a separate function (including basic error handling like _missing_ header and _invalid/unparsable date_)

File: be/src/services/vocabulary.service.ts
Comment (ja-fra) [inline be/src/services/vocabulary.service.ts]: there's a better way to do this: `$currentDate`

File: be/src/services/vocabulary.service.ts
Comment (ja-fra) [inline be/src/services/vocabulary.service.ts]: Check out the operation `findAndModify` which allows you combining the update and retrieval request in one query.

https://mongodb.github.io/node-mongodb-native/4.2/classes/Collection.html#findOneAndUpdate

File: be/src/services/vocabulary.service.ts
Comment (ja-fra) [inline be/src/services/vocabulary.service.ts]: check for `$eq`

**Review by ja-fra — CHANGES_REQUESTED** [review]

File: be/src/controllers/rest.router.ts
Comment (ManuelTS) [reply be/src/controllers/rest.router.ts]: > Since we will need this pattern at several places, please move it into a separate function (including basic error handling like _missing_ header and _invalid/unparsable date_)

Talked and implemented with @dan0bar on his branch, I will wait until he has merged and then use it in this branch.

File: be/src/services/vocabulary.service.ts
Comment (ManuelTS) [reply be/src/services/vocabulary.service.ts]: Will do, I used `lte` because of the `since` in `ifUnmodifiedSince`

File: be/src/services/vocabulary.service.ts
Comment (ja-fra) [reply be/src/services/vocabulary.service.ts]: If you're using the `$currentDate`, do not explicitly write it with `$set`.

**Review by ja-fra — CHANGES_REQUESTED** [review]

File: be/src/services/vocabulary.service.ts
Comment (ManuelTS) [reply be/src/services/vocabulary.service.ts]: > Check out the operation `findAndModify` which allows you combining the update and retrieval request in one query.
> 
> https://mongodb.github.io/node-mongodb-native/4.2/classes/Collection.html#findOneAndUpdate

@ja-fra With `$eq: ifUnmodifiedSince`set `findAndModify` will not find the entry if it has changed in the meantime. So I the capability of erroring with `412, 'Precondition Failed'` will be lost and hidden in `404, 'Not found'`. Is that ok?

Comment (ja-fra) [reply be/src/services/vocabulary.service.ts]: Nope. In the error-case you would need to check why it failed in a separate query: `404` or `412`

File: be/src/controllers/rest.router.ts
Comment (ja-fra) [inline be/src/controllers/rest.router.ts]: I think this should be `428`

File: be/src/services/vocabulary.service.ts
Comment (ja-fra) [inline be/src/services/vocabulary.service.ts]: Using `{ new: true }` will return you the updated/modified version from the database.

See https://docs.mongodb.com/manual/reference/method/db.collection.findAndModify/

**Review by ja-fra — CHANGES_REQUESTED** [review]

File: be/src/services/vocabulary.service.ts
Comment (ManuelTS) [reply be/src/services/vocabulary.service.ts]: For the protocol, the correct link for our mongoDB version is
https://docs.mongodb.com/manual/reference/method/db.collection.findOneAndUpdate/

**Comment by ManuelTS** [comment]
- [x] Checked with insomnia that PUT vocabs works
- [x] Checked in FE that PUT is used and that it works in FE too

**Review by dan0bar — APPROVED** [review]

**Pull request merged**
